### PR TITLE
feat: Add requests idempotency verification on the server side

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
The idea behind that is to be able to properly handle client side retries. 

How this is supposed to work:
- The client adds a randomly generated `X-Idempotency-Key` header to each request we'd like to make idempotent (each request must have a different value). The server itself only considers header presence in POST and PUT requests.
- The server temporarily stores the response stream of the marked request in the internal LRU cache. After the request is done then its result is serialised to the file system.
- If the second request with an idempotency key is coming and no matches are found then the request proceeds further
- If there is an active request with the same key then the response stream of that request will be rerouted to the current one, so the client would receive the same response
- If there was a request with the same key then the previously serialized cached response would be piped to the current response
- All cached responses are removed from the cache automatically as soon as they expire or the cache reaches its limit, since its a LRU cache type

Read https://stripe.com/docs/api/idempotent_requests for more details on this topic

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
This change will require to tune the clients, although I think the improvement worth it

**Comment by mykola-mokhnach** [comment]
cc @ki4070ma

**Comment by jlipps** [comment]
Can you first help me understand in which scenarios we want client side retries?

**Comment by mykola-mokhnach** [comment]
@jlipps We don't control it. Client libraries, like OkHttp (the one that java client uses), apply retries automatically. See https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/-builder/retry-on-connection-failure/ for more details

File: package.json [hunk 1]
```diff
@@ -53,7 +53,7 @@
     "serve-favicon": "^2.4.5",
     "source-map-support": "^0.5.5",
     "validate.js": "^0.13.0",
-    "webdriverio": "^5.10.9",
+    "webdriverio": "^6.0.2",
```
Comment (imurchie) [inline package.json]: Why is this bumped in _this_ PR? Seems unrelated.

File: lib/express/idempotency.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+import os from 'os';
+import path from 'path';
+import { EventEmitter } from 'events';
+
+
+const CACHE_SIZE = 1024;
+const IDEMPOTENT_RESPONSES = new LRU({
+  max: CACHE_SIZE,
+  updateAgeOnGet: true,
+  dispose: (key, {response}) => {
```
Comment (imurchie) [inline lib/express/idempotency.js]: No need to make an anonymous function here 
```js
const IDEMPOTENT_RESPONSES = new LRU({
  max: CACHE_SIZE,
  updateAgeOnGet: true,
  dispose (key, {response}) {
    if (response) {
      fs.rimrafSync(response);
    }
  },
});
```

File: package.json (hunk 1 above)
Comment (mykola-mokhnach) [reply package.json]: rebase glitch. Will revert

**Comment by KazuCocoa** [comment]
👍
